### PR TITLE
Check for duplicate endpoint names on startup

### LINKED ISSUE #35436: Fix minimal APIs behavior with duplicate endpoint and route names
By default, endpoint names must be global within an application.

MVC currently checks for duplicate endpoint names in the `EndpointNameAddressScheme` which maintains a cache that when hit, will run through all the declared endpoint names and throw if any errors are encountered.

https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/src/EndpointNameAddressScheme.cs#L38-L41

So, if a user has an application with duplicate endpoint names invoked, the application will through when one of the endpoints is hit (not at startup because the check only happens when the cache is initialized).

The same validations don't happen in minimal APIs because we don't execute any endpoints that invoke the cache initialization path.

### PR DESCRIPTION
**Pull request by captainsafia**
Fixes https://github.com/dotnet/aspnetcore/issues/35436

### CODE REVIEW

File: src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs [hunk 1]
```diff
@@ -41,13 +41,25 @@ public override Task MatchAsync(HttpContext httpContext)
         private Matcher CreateMatcher(IReadOnlyList<Endpoint> endpoints)
         {
             var builder = _matcherBuilderFactory();
+            var seenEndpointNames = new List<string>();
```
Comment (davidfowl) [inline src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs:44]:
```suggestion
            var seenEndpointNames = new HashSet<string>();
```

File: src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs [hunk 2]
```diff
@@ -41,13 +41,25 @@ public override Task MatchAsync(HttpContext httpContext)
...
                 // By design we only look at RouteEndpoint here. It's possible to
                 // register other endpoint types, which are non-routable, and it's
                 // ok that we won't route to them.
                 if (endpoints[i] is RouteEndpoint endpoint && endpoint.Metadata.GetMetadata<ISuppressMatchingMetadata>()?.SuppressMatching != true)
                 {
+                    // Validate that endpoint names are unique.
+                    var endpointName = endpoint.Metadata.GetMetadata<IEndpointNameMetadata>()?.EndpointName;
+                    if (endpointName != null)
+                    {
+                        if (seenEndpointNames.Contains(endpointName))
```
Comment (davidfowl) [inline src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs:56]:
```suggestion
                        if (seenEndpointNames.Add(endpointName))
```

File: src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs [hunk 3]
```diff
@@ -41,13 +41,25 @@ public override Task MatchAsync(HttpContext httpContext)
...
+                        {
+                            throw new Exception($"Duplicate endpoint name {endpointName} found. Endpoint names must be globally unique.");
+                        }
+                        seenEndpointNames.Add(endpointName);
```
Comment (davidfowl) [inline src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs:60]:
```suggestion
```

File: src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs (hunk 1 above)
Comment (davidfowl) [reply src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs:44]: Case-insensitive?

File: src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs [hunk 4]
```diff
@@ -41,13 +41,25 @@ public override Task MatchAsync(HttpContext httpContext)
...
             for (var i = 0; i < endpoints.Count; i++)
             {
                 // By design we only look at RouteEndpoint here. It's possible to
                 // register other endpoint types, which are non-routable, and it's
                 // ok that we won't route to them.
                 if (endpoints[i] is RouteEndpoint endpoint && endpoint.Metadata.GetMetadata<ISuppressMatchingMetadata>()?.SuppressMatching != true)
                 {
+                    // Validate that endpoint names are unique.
+                    var endpointName = endpoint.Metadata.GetMetadata<IEndpointNameMetadata>()?.EndpointName;
+                    if (endpointName != null)
```
Comment (davidfowl) [inline src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs:54]:
```suggestion
                    if (endpointName is not null)
```

File: src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs [hunk 5]
```diff
@@ -41,13 +41,25 @@ public override Task MatchAsync(HttpContext httpContext)
...
+                    {
+                        if (seenEndpointNames.Contains(endpointName))
+                        {
+                            throw new Exception($"Duplicate endpoint name {endpointName} found. Endpoint names must be globally unique.");
```
Comment (davidfowl) [inline src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs:58]: We never throw the base Exception, maybe InvalidOperationException.

Comment (captainsafia) [reply src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs:58]: Yeah, `InvalidOperationException` didn't seem quite right either because it's moreso invalid data. I was trying to find an exception for that...

Comment (captainsafia) [reply src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs:58]: Actually, maybe `InvalidOperationException` is fine...

File: src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs [hunk 6]
```diff
@@ -41,13 +41,24 @@ public override Task MatchAsync(HttpContext httpContext)
...
                 // ok that we won't route to them.
                 if (endpoints[i] is RouteEndpoint endpoint && endpoint.Metadata.GetMetadata<ISuppressMatchingMetadata>()?.SuppressMatching != true)
                 {
+                    // Validate that endpoint names are unique.
+                    var endpointName = endpoint.Metadata.GetMetadata<IEndpointNameMetadata>()?.EndpointName;
+                    if (endpointName is not null)
+                    {
+                        if (!seenEndpointNames.Add(endpointName))
+                        {
+                            throw new InvalidOperationException($"Duplicate endpoint name {endpointName} found. Endpoint names must be globally unique.");
```
[the remark this answers is not in the document]
Comment (captainsafia) [reply src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs:58]: Yeah, I think it'll be a little bit helpful for folks trying to debug scenarios where the duplicate endpoint name is added automagically.

File: src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs (hunk 1 above)
Comment (captainsafia) [reply src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs]: Looks like they are case sensitive. The other checks don't ignore casing:

https://github.com/dotnet/aspnetcore/blob/e23fd047b4baf3480ce93d7643a897732f960557/src/Http/Routing/src/EndpointNameAddressScheme.cs#L53-L59

File: src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Kahbazi) [reply src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs:58]: In that case we might want to use `Dictionary<string, Endpoint>` instead of `HashSet<string>` so we could have the `DisplayName` of both endpoint.

File: src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs [hunk 7]
```diff
@@ -41,13 +41,24 @@ public override Task MatchAsync(HttpContext httpContext)
...
                 // register other endpoint types, which are non-routable, and it's
                 // ok that we won't route to them.
                 if (endpoints[i] is RouteEndpoint endpoint && endpoint.Metadata.GetMetadata<ISuppressMatchingMetadata>()?.SuppressMatching != true)
                 {
+                    // Validate that endpoint names are unique.
+                    var endpointName = endpoint.Metadata.GetMetadata<IEndpointNameMetadata>()?.EndpointName;
+                    if (endpointName is not null)
+                    {
+                        if (!seenEndpointNames.Add(endpointName))
+                        {
```
Comment (Kahbazi) [inline src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs:57]: Does this count as a breaking change!? Should there be an AppContext switch to skip this check?

**Comment by captainsafia** [comment]
> Does this count as a breaking change!? Should there be an AppContext switch to skip this check?

This is technically not a breaking change. The policy has always been that duplicate endpoint names are not allowed in an application. The subtlety is that in MVC, route names map to endpoint names and we already have logic that checks to see if duplicate route names exist. There's also pre-existing logic  in MVC that checks for duplicate endpoint names and throws an IOE. That logic was being superseded in the route name check.

https://github.com/dotnet/aspnetcore/blob/c23a0e6f5d12d1f64412cfcda1428f08d860719c/src/Http/Routing/src/EndpointNameAddressScheme.cs#L92

So, TLDR, we are throwing this exception in a new place but it's not a breaking change in the sense that it is not introducing a new assumption.

**Review by javiercn — APPROVED** [review]
Looks great!

**Review by BrennanConroy — APPROVED** [review]
